core/templates/pages/collection-player-page/collection-player-page.component.ts

```diff
@@ -0,0 +1,433 @@
+      }
+    }
+  }
+
+  generatePathIconParameters(): IconParametersArray[] {
+    let collectionNodes = this.collection.getCollectionNodes();
+    let iconParametersArray = [];
+    iconParametersArray.push({
+      thumbnailIconUrl:
+        collectionNodes[0].getExplorationSummaryObject(
+        ).thumbnail_icon_url.replace('subjects', 'inverted_subjects'),
+      left: '225px',
+      top: '35px',
+      thumbnailBgColor:
+        collectionNodes[0].getExplorationSummaryObject(
+        ).thumbnail_bg_color
+    });
+
+    // Here x and y represent the co-ordinates for the icons in the
+    // path.
+    let x = this.ICON_X_MIDDLE_PX;
+    let y = this.ICON_Y_INITIAL_PX;
+    let countMiddleIcon = 1;
+
+    for (let i = 1; i < this.collection.getCollectionNodeCount(); i++) {
+      if (countMiddleIcon === 0 && x === this.ICON_X_MIDDLE_PX) {
+        x = this.ICON_X_LEFT_PX;
+        y += this.ICON_Y_INCREMENT_PX;
+        countMiddleIcon = 1;
+      } else if (countMiddleIcon === 1 && x === this.ICON_X_MIDDLE_PX) {
+        x = this.ICON_X_RIGHT_PX;
+        y += this.ICON_Y_INCREMENT_PX;
+        countMiddleIcon = 0;
+      } else {
+        x = this.ICON_X_MIDDLE_PX;
+        y += this.ICON_Y_INCREMENT_PX;
+      }
+      iconParametersArray.push({
+        thumbnailIconUrl:
+          collectionNodes[i].getExplorationSummaryObject(
+          ).thumbnail_icon_url.replace(
+            'subjects', 'inverted_subjects'),
+        left: x + 'px',
+        top: y + 'px',
+        thumbnailBgColor:
+          collectionNodes[i].getExplorationSummaryObject(
+          ).thumbnail_bg_color
+      });
+    }
+    return iconParametersArray;
+  }
+
+  getExplorationUrl(explorationId: string): string {
+    return (
+      '/explore/' + explorationId + '?collection_id=' +
+      this.collectionId);
+  }
+
+  getExplorationTitlePosition(index: number): string {
+    if (index % 2 === 0) {
+      return '8px';
+    } else if ((index + 1) % 2 === 0 && (index + 1) % 4 !== 0) {
+      return '30px';
+    } else if ((index + 1) % 4 === 0) {
+      return '-40px';
+    }
+  }
+
+  scrollToLocation(id: string): void {
+    this.windowRef.nativeWindow.location.hash = '#' + (id);
+  }
+
+  closeOnClickingOutside(): void {
+    this.explorationCardIsShown = false;
+  }
+
+  onClickStopPropagation($evt: Event): void {
+    $evt.stopPropagation();
+  }
+
+  isCompletedExploration(explorationId: string): boolean {
+    if (this.collectionPlaythrough) {
+      let completedExplorationIds = (
+        this.collectionPlaythrough.getCompletedExplorationIds());
+      return completedExplorationIds.indexOf(explorationId) > -1;
+    }
+    return false;
+  }
+
+  ngOnInit(): void {
+    // $scope.$watch('$ctrl.collection', function(newValue) {
+    //   if (
+    //     newValue !== null &&
+    //     ctrl.collection.getCollectionNodeCount()) {
+    //     ctrl.generatePathParameters();
+    //   }
+    // }, true);
```

Comment:
Suggestion:

Never directly change the value of `$ctrl.collection` but instead create a function that assign the new value
and also run this code in the callback thereafter.

Reply:
I think `ngOnChanges` would be more appropriate here ?

Reply:
I don't agree, this pattern runs code even if some other variable instead of collection is changed. Even if go with this pattern, you have add additional if statement checking whether collection is changed (for that another variable holding previous value is required).

The way we have handled this case in the rest of codebase (angular). We can create event emitter (eg. onCollectionChanged) to notify the subscribers that the value is changed.

Here that won't be required as the collection variable is only changed once in the code.

The if statement and call to generatePathParameters (from scope.watch) can be made just after assignment of this.collection = ....

To make it more re usable, the assignment and code from scope watch can be put into a single function called updateCollection. Which is not required unless collection is assigned more then once in the component.

1. The function updateCollection design is a reusable and tightly coupled approach. Hence, more interdependence of assignment and code from scope.watch
2. The event emitter approach is reusable and loosely coupled. Which results in less interdependence.